### PR TITLE
feat(store): add getFieldLength and getFieldSlice

### PR DESCRIPTION
**Pull request by dk1a**
fixes https://github.com/latticexyz/mud/issues/735
fixes https://github.com/latticexyz/mud/issues/736

World size changed less than I expected: 21.564 -> 22.167

### CODE REVIEW

File: packages/common/src/codegen/render-solidity/renderTypeHelpers.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export function renderTypeHelpers(options: { fields: RenderField[]; primaryKeys:
   }
 
   // bool is special - it's the only primitive value type that can't be typecasted to/from
-  if (fields.some(({ typeId }) => typeId === "bool")) {
+  if (fields.some(({ internalTypeId }) => internalTypeId.match("bool"))) {
```
Comment (alvrs) [inline packages/common/src/codegen/render-solidity/renderTypeHelpers.ts:13]: for context - why this change?

File: packages/store/src/StoreCore.sol [hunk 2]
```diff
@@ -476,6 +476,50 @@ library StoreCore {
...
+   * Get a byte slice of a single dynamic field from the given tableId and key tuple, with the given schema
+   * The slice is unchecked and will return invalid data if `start`:`end` overflow
+   */
+  function getFieldSlice(
+    bytes32 tableId,
+    bytes32[] memory key,
+    uint8 schemaIndex,
+    Schema schema,
+    uint256 start,
+    uint256 end
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/StoreCore.sol:510]: ooc what's the rationale behind start/end vs start/length?

File: packages/common/src/codegen/render-solidity/renderTypeHelpers.ts (hunk 1 above)
Comment (dk1a) [reply packages/common/src/codegen/render-solidity/renderTypeHelpers.ts:13]: `.match` - because arrays may need their items converted, and this is simpler than `["bool", "bool[]"].includes` and might include something else I forgot
`internal` - if I add more userTypes a bool could be wrapped, so typeId wouldn't help here

**Review by alvrs — APPROVED** [review]
lgtm!

**Review by alvrs — APPROVED** [review]

**Review by alvrs — APPROVED** [review]

File: packages/store/src/StoreCore.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/StoreCore.sol]: mostly consistency: js, solidity (it can only slice calldata though), our `Slice`, and rust all use `start` and `end` (they similarly include start and exclude end, which is why I took it for granted and didn't mention)
imo when working with indexes it's more convenient/clear
